File: sdk/cpp/tests/integration_test.cpp
#include "libravdb.hpp"
#include <iostream>
#include <cassert>
#include <filesystem>

using namespace libravdb;
namespace fs = std::filesystem;

int main() {
    std::string db_path = "./demo_db_cpp";
    if (fs::exists(db_path)) {
        fs::remove_all(db_path);
    }

    try {
        std::cout << "Initializing LibraVDB at " << db_path << "..." << std::endl;
        LibraVDB db(db_path);

        std::cout << "Database Ping OK." << std::endl;
        db.ping();

        std::cout << "Setting global memory limit..." << std::endl;
        db.set_memory_limit(10 * 1024 * 1024);

        std::cout << "Creating collection 'docs' (dim: 3)..." << std::endl;
        Collection col = db.create_collection("docs", 3);

        auto collections = db.list_collections();
        std::cout << "List collections: [" << (collections.empty() ? "" : collections[0]) << "]" << std::endl;
        assert(collections.size() == 1 && collections[0] == "docs");

        std::cout << "Testing InsertBatch with 1000 vectors..." << std::endl;
        std::vector<std::string> ids;
        std::vector<std::vector<float>> vectors;
        std::vector<json> metadata;

        for (int i = 0; i < 1000; ++i) {
            ids.push_back("vec_" + std::to_string(i));
            vectors.push_back({0.1f, 0.2f, 0.3f});
            metadata.push_back({
                {"source", "cpp_test"},
                {"index", i},
                {"active", (i % 2 == 0)}
            });
        }

        col.insert_batch(ids, vectors, metadata);
        std::cout << "Batch Insert Complete!" << std::endl;

        std::cout << "Testing Update..." << std::endl;
        col.update("vec_0", {1.0f, 1.0f, 1.0f}, json{{"updated", true}});

        std::cout << "Testing Get..." << std::endl;
        json rec = col.get("vec_0");
        std::cout << "Got record: " << rec.dump() << std::endl;
        assert(rec["id"] == "vec_0");

        std::cout << "Testing Search with Filters (active = true)..." << std::endl;
        Filter filter = Filter::eq("active", true);
        json results = col.search({1.0f, 1.0f, 1.0f}, 5, filter);
        std::cout << "Search Results: " << results.dump() << std::endl;
        assert(results.is_array() && !results.empty());

        std::cout << "Testing Scan (Offset 0, Limit 2)..." << std::endl;
        json scanned = col.scan(0, 2);
        std::cout << "Scanned: " << scanned.dump() << std::endl;
        assert(scanned.is_array() && scanned.size() == 2);

        std::cout << "Testing DeleteBatch (Deleting 500 vectors)..." << std::endl;
        std::vector<std::string> del_ids(ids.begin(), ids.begin() + 500);
        col.delete_batch(del_ids);

        int64_t count = col.count();
        std::cout << "Collection Count: " << count << std::endl;
        assert(count == 500);

        std::cout << "Testing Vacuum..." << std::endl;
        db.vacuum();

        std::cout << "Dropping database..." << std::endl;
        db.drop_db();
        
        assert(!fs::exists(db_path));
        std::cout << "Database successfully dropped!" << std::endl;

    } catch (const LibraException& e) {
        std::cerr << "Test failed with error: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}

File: sdk/rust/tests/integration_test.rs
use libravdb::{LibraVDB, Filter};
use serde_json::json;
use std::fs;

#[test]
fn test_libravdb_integration() {
    let db_path = "./demo_db_rust";
    if fs::metadata(db_path).is_ok() {
        fs::remove_dir_all(db_path).unwrap();
    }

    println!("Initializing LibraVDB at {}...", db_path);
    let db = LibraVDB::new(db_path).expect("Failed to open DB");

    println!("Database Ping OK: {:?}", db.ping());

    println!("Setting global memory limit...");
    db.set_memory_limit(10 * 1024 * 1024).unwrap();

    println!("Creating collection 'docs' (dim: 3)...");
    let col = db.create_collection("docs", 3).unwrap();

    let collections = db.list_collections().unwrap();
    println!("List collections: {:?}", collections);
    assert_eq!(collections, vec!["docs"]);

    println!("Testing InsertBatch with 1000 vectors...");
    let mut ids = Vec::new();
    let mut vectors = Vec::new();
    let mut metadata = Vec::new();

    for i in 0..1000 {
        ids.push(format!("vec_{}", i));
        vectors.push(vec![0.1_f32, 0.2_f32, 0.3_f32]);
        metadata.push(json!({
            "source": "rust_test",
            "index": i,
            "active": i % 2 == 0
        }));
    }

    col.insert_batch(&ids, &vectors, Some(&metadata)).unwrap();
    println!("Batch Insert Complete!");

    println!("Testing Update...");
    col.update("vec_0", &[1.0, 1.0, 1.0], Some(json!({ "updated": true }))).unwrap();

    println!("Testing Get...");
    let rec = col.get("vec_0").unwrap();
    println!("Got record: {}", rec);
    assert_eq!(rec["id"], "vec_0");

    println!("Testing Search with Filters (active = true)...");
    let filter = Filter::eq("active", true);
    let results = col.search(&[1.0, 1.0, 1.0], 5, Some(&filter)).unwrap();
    println!("Search Results: {:?}", results);
    assert!(results.as_array().unwrap().len() > 0);

    println!("Testing Scan (Offset 0, Limit 2)...");
    let scanned = col.scan(0, 2).unwrap();
    println!("Scanned: {}", scanned);
    assert_eq!(scanned.as_array().unwrap().len(), 2);

    println!("Testing DeleteBatch (Deleting 500 vectors)...");
    col.delete_batch(&ids[0..500]).unwrap();

    let count = col.count().unwrap();
    println!("Collection Count: {}", count);
    assert_eq!(count, 500);

    println!("Testing Vacuum...");
    db.vacuum().unwrap();

    println!("Dropping database...");
    db.drop_db().unwrap();

    assert!(fs::metadata(db_path).is_err());
    println!("Database successfully dropped!");
}
